domain/machine/state/state.go

```diff
@@ -857,3 +858,108 @@ WHERE  name = $machineName.name`
 		return nil
 	})
 }
+
+// LXDProfiles returns the names of the LXD profiles on the machine.
+func (st *State) LXDProfiles(ctx context.Context, mUUID string) ([]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	machineUUIDQuery := machineUUID{UUID: mUUID}
+	query := `
+SELECT   &lxdProfile.name
+FROM     machine_lxd_profile
+WHERE    machine_uuid = $machineUUID.uuid
+ORDER BY array_index ASC`
+	queryStmt, err := st.Prepare(query, lxdProfile{}, machineUUIDQuery)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var result []lxdProfile
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, queryStmt, machineUUIDQuery).GetAll(&result)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return internalerrors.Errorf("retrieving lxd profiles for machine %q: %w", mUUID, err)
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	lxdProfiles := make([]string, len(result))
+	for i, lxdProfile := range result {
+		lxdProfiles[i] = lxdProfile.Name
+	}
+	return lxdProfiles, nil
+}
+
+// SetLXDProfiles sets the list of LXD profile names to the lxd_profile table
+// for the given machine. This method will overwrite the list of profiles for
+// the given machine without any checks.
+func (st *State) SetLXDProfiles(ctx context.Context, mUUID string, profileNames []string) error {
+	if len(profileNames) == 0 {
+		return nil
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return internalerrors.Errorf("cannot get database to set lxd profiles %v for machine %q: %w", profileNames, mUUID, err)
+	}
+
+	checkMachineExistsStmt, err := st.Prepare(`
+SELECT uuid AS &machineUUID.uuid
+FROM   machine
+WHERE  machine.uuid = $machineUUID.uuid`, machineUUID{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	removePreviousProfilesStmt, err := st.Prepare(
+		`DELETE FROM machine_lxd_profile WHERE machine_uuid = $machineUUID.uuid`,
+		machineUUID{},
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	setLXDProfileStmt, err := st.Prepare(`
+INSERT INTO machine_lxd_profile (*)
+VALUES      ($lxdProfile.*)`, lxdProfile{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		queryMachineUUID := machineUUID{UUID: mUUID}
+		var machineExists machineUUID
+		err = tx.Query(ctx, checkMachineExistsStmt, queryMachineUUID).Get(&machineExists)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return machineerrors.MachineNotFound
+		} else if err != nil {
+			return internalerrors.Errorf("checking if machine %q exists: %w", mUUID, err)
+		}
+
+		// Make sure to clean up any existing profiles on the given machine.
```

Comment:
Before deleting the profiles, can we check that the profiles in the rows don't already match the ones that already exist. It's very common for the SetLXDProfiles to be triggered with the same content. This comes back to only do a write query if you _really_ need to do one.

When we do wire up lxd profiles, we'll need to insert a logical row into the change stream at the end of this to fire the worker.

Reply:
Done, see 618929544d.

I'm interested in yours and @manadart 's thoughts on what I did, because this includes now some (small amount of) logic in the state layer. This case sits on the edge of where I'd start using RunAtomic, but IMO it can still live in the state layer.

Reply:
Keep it here for now.